src/dp/distributions.rs

```diff
@@ -0,0 +1,340 @@
+        let y_abs: Ratio<BigUint> = BigUint::new(y.to_u32_digits().1).into();
+
+        let sub = sigma.pow(2) / t.clone();
+        let fact = (sigma.pow(2) * BigUint::from(2u8)).recip();
+
+        let prob: Ratio<BigUint> = if y_abs < sub {
+            (sub - y_abs).pow(2) * fact
+        } else {
+            (y_abs - sub).pow(2) * fact
+        };
+        if sample_bernoulli_exp(&prob, rng) {
+            return y;
+        }
+    }
+}
+
+/// samples `BigInt` numbers according to the discrete Gaussian distribution with mean zero.
+/// The distribution is defined over the integers, represented by arbitrary-precision integers.
+/// The sampling procedure follows [[CKS20]].
+///
+/// [CKS20]: https://arxiv.org/abs/2004.00010
+#[derive(Clone, Debug)]
+pub struct DiscreteGaussian {
+    /// The standard deviation of the distribution.
+    std: Ratio<BigUint>,
+}
+
+impl DiscreteGaussian {
+    /// Create a new sampler from the Discrete Gaussian Distribution with the given
+    /// standard deviation and mean zero. Errors if the input has denominator zero.
+    pub fn new_checked(std: Ratio<BigUint>) -> Result<DiscreteGaussian, DpError> {
+        if std.denom().is_zero() {
+            return Err(DpError::ZeroDenominator());
+        }
+        Ok(DiscreteGaussian { std })
+    }
+}
+
+impl Distribution<BigInt> for DiscreteGaussian {
+    fn sample<R>(&self, rng: &mut R) -> BigInt
+    where
+        R: Rng + ?Sized,
+    {
+        sample_discrete_gaussian(&self.std, rng)
+    }
+}
+
+impl DifferentialPrivacyDistribution for DiscreteGaussian {}
+
+/// A DP strategy using the discrete gaussian distribution.
+pub struct DiscreteGaussianDpStrategy<B>
+where
+    B: DifferentialPrivacyBudget,
+{
+    budget: B,
+}
+
+/// A DP strategy using the discrete gaussian distribution providing zero-concentrated DP.
+pub type ZCdpDiscreteGaussian = DiscreteGaussianDpStrategy<ZCdpBudget>;
+
+impl DifferentialPrivacyStrategy for DiscreteGaussianDpStrategy<ZCdpBudget> {
+    type Budget = ZCdpBudget;
+    type Distribution = DiscreteGaussian;
+    type Sensitivity = Ratio<BigUint>;
+
+    fn from_budget(budget: ZCdpBudget) -> DiscreteGaussianDpStrategy<ZCdpBudget> {
+        DiscreteGaussianDpStrategy { budget }
+    }
+
+    /// Create a new sampler from the Discrete Gaussian Distribution with a standard
+    /// deviation calibrated to provide `1/2 epsilon^2` zero-concentrated differential
+    /// privacy when added to the result of an integer-valued function with sensitivity
+    /// `sensitivity`, following Theorem 4 from [[CKS20]]
+    ///
+    /// [CKS20]: https://arxiv.org/pdf/2004.00010.pdf
+    fn create_distribution(
+        &self,
+        sensitivity: Ratio<BigUint>,
+    ) -> Result<DiscreteGaussian, DpError> {
+        match DiscreteGaussian::new_checked(sensitivity / self.budget.epsilon.clone()) {
+            Ok(d) => Ok(d),
+            Err(e) => Err(e),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
I think I saw some (commented out) statistical tests in a previous version of this PR. Did you have that?

 It could be very useful to have a kind of sanity check showing that we are not too far off from the right distribution, at least with high probability.

Reply:
i removed that because it felt weird to have a test that can, in some very unlikely cases, fail...but now that there's seeded randomness i can put them back (even tho they are only necessary and very not sufficient for it to be normal noise). is it ok to add `statrs` and `statest` as dev dependencies?

Reply:
I agree that flaky tests aren't really suitable for unit testing. I think they would be helpful to have in `examples`, however. (See the root directory of this project.)

Reply:
In general adding dev dependencies is not a huge problem (we just don't want the core crate features to have too many dependencies I think.)

Reply:
To close the loop on this: #648 added cargo-vet audits and exemptions to allow adding `statrs` and `statest` as development dependencies.